src/main/java/com/mraof/minestuck/blockentity/machine/AnthvilBlockEntity.java

```diff
@@ -0,0 +1,259 @@
+		return goodOnFuel || !input.isEmpty();
+	}
+	
+	/**
+	 * Only handles refreshing uranium
+	 */
+	private void processContents()
+	{
+		if(canBeRefueled())
+		{
+			//checks for a uranium itemstack in the right item slot, increases the fuel value if some is found and then removes one count from the fuel stack
+			if(itemHandler.getStackInSlot(1).is(ExtraForgeTags.Items.URANIUM_CHUNKS))
+			{
+				//Refill fuel
+				addFuel((short) FUEL_INCREASE);
+				itemHandler.extractItem(1, 1, false);
+			}
+		}
+	}
+	
+	public static void attemptMend(AnthvilBlockEntity anthvil, ServerPlayer player)
+	{
+		Level level = anthvil.level;
+		ItemStack slotStack = anthvil.itemHandler.getStackInSlot(0);
+		GristCache playerCache = GristCache.get(player);
+		
+		ImmutableGristSet pickedGrist = mendingGrist(anthvil, playerCache, level, slotStack);
+		
+		if(!GristAmount.EMPTY.equals(pickedGrist))
+		{
+			if(slotStack.hasCraftingRemainingItem())
+			{
+				anthvil.itemHandler.setStackInSlot(0, slotStack.getCraftingRemainingItem());
+			} else
+			{
+				//amount of repair changes with the value of the grist, 10 at base
+				int repairAmount = (int) (10 * pickedGrist.asAmounts().get(0).getValue());
+				slotStack.setDamageValue(slotStack.getDamageValue() - repairAmount);
+				
+				if(!player.isCreative() && playerCache.tryTake(pickedGrist, GristHelper.EnumSource.CLIENT))
+				{
+					anthvil.fuel -= MEND_FUEL_COST;
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Checks that there is enough fuel energy/grist for the machine to work and that there is something valid to mend, returns the appropriate grist amount if so
+	 */
+	private static ImmutableGristSet mendingGrist(AnthvilBlockEntity anthvil, GristCache playerCache, Level level, ItemStack slotStack)
+	{
+		if(level != null && !slotStack.isEmpty() && slotStack.isRepairable() && slotStack.isDamageableItem() && slotStack.isDamaged())
+		{
+			GristSet fullSet = GristCostRecipe.findCostForItem(slotStack, null, false, level);
+			
+			if(fullSet != null && !fullSet.isEmpty())
+			{
+				GristAmount pickedGrist = getUsedGrist(fullSet);
+				
+				if(anthvil.fuel >= MEND_FUEL_COST && playerCache.canAfford(pickedGrist))
+					return pickedGrist;
+			}
+		}
+		
+		return GristAmount.EMPTY;
+	}
+	
+	/**
+	 * Takes the GristSet of the item stored in the mending slot of the anthvil, finds the most used grist type, then returns a GristSet composed of one value of said grist type
+	 */
+	public static GristAmount getUsedGrist(GristSet fullSet)
+	{
+		List<GristAmount> gristAmounts = fullSet.asAmounts();
+		GristAmount pickedGrist = new GristAmount(GristTypes.BUILD.get(), 1);
+		
+		if(gristAmounts.size() > 1) //establishes which GristAmount in the set has the highest impact and uses that
+		{
+			for(GristAmount grist : fullSet.asAmounts())
+			{
+				double gristValue = grist.getValue();
+				if(grist.type() == GristTypes.BUILD.get())
+					gristValue /= 10; //reduces build grist value
```

Comment:
Make a local/private function for this modified value and you could use a stream:
`fullSet.asAmounts().stream().max(Comparator.comparingDouble(AnthvilBlockEntity::modifiedGristValue))`
(this expression could then replace essentially the entirety of `getUsedGrist()`)

Reply:
Oh actually, that wouldn't be true for items with a negative grist cost, but you can be the judge of that.